internal/metrics/directory_usage_collector.go

```diff
@@ -0,0 +1,69 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	directoryUsageUsedBytesName        = "assisted_installer_directory_usage_bytes"
+	directoryUsageUsedBytesDescription = "The amount of the free space in a directory that has been used in bytes"
+	fsUsageFreeBytesName               = "assisted_installer_fs_free_bytes"
+	fsUsageFreeBytesDescription        = "The amount of free space that is remaining for the filesystem backing this directory"
+
+	directoryLabelKey = "directory"
+)
+
+type directoryUsageCollectorConfig struct {
+	directories []string
+}
+
+func NewDirectoryUsageCollectorConfig(directories []string) *directoryUsageCollectorConfig {
+	return &directoryUsageCollectorConfig{directories: directories}
+}
+
+type directoryUsageCollector struct {
+	// The directory for which to report disk usage stats
+	directories              []string
+	diskStatsHelper          DiskStatsHelper
+	directoryUsageBytesDescs map[string]*prometheus.Desc
+	fsFreeByteDescs          map[string]*prometheus.Desc
+	log                      *logrus.Logger
+}
+
+func newDirectoryUsageCollector(directories []string, diskStatsHelper DiskStatsHelper, log *logrus.Logger) *directoryUsageCollector {
+	directoryUsageBytesDescs := make(map[string]*prometheus.Desc)
+	fsFreeByteDescs := make(map[string]*prometheus.Desc)
+	for _, directory := range directories {
+		directoryUsageBytesDescs[directory] = prometheus.NewDesc(directoryUsageUsedBytesName, directoryUsageUsedBytesDescription, nil, map[string]string{directoryLabelKey: directory})
+		fsFreeByteDescs[directory] = prometheus.NewDesc(fsUsageFreeBytesName, fsUsageFreeBytesDescription, nil, map[string]string{directoryLabelKey: directory})
+	}
+	return &directoryUsageCollector{
+		directories:              directories,
+		diskStatsHelper:          diskStatsHelper,
+		directoryUsageBytesDescs: directoryUsageBytesDescs,
+		fsFreeByteDescs:          fsFreeByteDescs,
+		log:                      log,
+	}
+}
+
+func (c *directoryUsageCollector) Describe(ch chan<- *prometheus.Desc) {
+	for _, desc := range c.directoryUsageBytesDescs {
+		ch <- desc
+	}
+	for _, desc := range c.fsFreeByteDescs {
+		ch <- desc
+	}
+}
+
+func (c *directoryUsageCollector) Collect(ch chan<- prometheus.Metric) {
+	for _, directory := range c.directories {
```

Comment:
As `directories`  is a map we shouldn't rely on its order to do this, as golang randomizes access.

Reply:
The order here is not actually important, only that each directory is handled.

I don't see this as a problem in this case

Reply:
How can each metric be associated with its description? Isn't that based on order?

Reply:
It's part of the map, directory is the key and there are two maps,

One for the descs regarding disk usage
One for the descs regarding free space

In both cases, looking up the directory by key will give you the correct desc for that directory.